### PR TITLE
Flatten Activity types

### PR DESCRIPTION
**Pull request by jenniferplusplus**
I'm trying out APSharp again for some more complex activities. It's handling things pretty well, overall. There are a couple failures here, but honestly those are not showstoppers right now.

This includes some fixes for Activity properties that were missing, or were Linkable<T>, but should have been LinkableList<T>. I'm not sure what to do about the language-mapped strings. I think the most interoperable thing will be to have both the language-mapped and language-insensitive properties in the model.

### CODE REVIEW

File: Source/ActivityPub.Types/AS/ASActivity.cs [hunk 1]
```diff
@@ -92,11 +103,23 @@ public LinkableList<ASObject> Actor
...
         set => Entity.Result = value;
     }
+
+    /// <summary>
+    /// Describes the indirect object, or target, of the activity. The precise meaning of the target is largely
+    /// dependent on the type of action being described but will often be the object of the English preposition "to".
+    /// For instance, in the activity "John added a movie to his wishlist", the target of the activity is John's
+    /// wishlist. An activity can have more than one target. 
+    /// </summary>
+    public LinkableList<ASObject>? Target
```
Comment (warriordog) [inline Source/ActivityPub.Types/AS/ASActivity.cs:118]: this one exists in `ASTargetedActivity` for the same reason as `Object`. It's somewhat convoluted, but this protects against null objects or targets when they might be required. It also ensures that they *are* null for those activity types that prohibit them.

**Review by warriordog — CHANGES_REQUESTED** [review]
The missing properties are not actually missing - they're just in a subclass. I'm open to changing this if you think we should. Thoughts?

**Comment by jenniferplusplus** [comment]
@warriordog I see, I'll have another look. I added the Object property because it wasn't available on one of the types I was expecting. Same with Target. I forget which, so I'll have to go back and see. The transitive/intransitive split makes sense, in that it's in keeping with the spec as written. Whether it works in practice is a question of whether other implementations also implement that properly, and tbh I don't know if they do or not.

For Target, the AS2 spec says it's a property of Activity, so I would expect it to be available on every activity.

**Comment by warriordog** [comment]
@jenniferplusplus I'm open to making them base-level properties if you think that's better. We could remove the transitive / targeted activities to match.

File: Source/ActivityPub.Types/AS/ASActivity.cs [hunk 2]
```diff
@@ -69,19 +69,30 @@ public LinkableList<ASObject> Actor
     ///     Identifies one or more objects used (or to be used) in the completion of an Activity.
     /// </summary>
     /// <seealso href="https://www.w3.org/TR/activitystreams-vocabulary/#dfn-instrument" />
-    public Linkable<ASObject>? Instrument
+    public LinkableList<ASObject>? Instrument
```
[the remark this answers is not in the document]
Comment (warriordog) [reply Source/ActivityPub.Types/AS/ASActivity.cs:72]: nevermind, apparently I have _not_ been doing that.

File: Source/ActivityPub.Types/AS/ASActivity.cs [hunk 3]
```diff
@@ -69,19 +69,30 @@ public LinkableList<ASObject> Actor
...
         get => Entity.Instrument;
         set => Entity.Instrument = value;
     }
 
+    /// <summary>
+    /// Describes an object of any kind. The Object type serves as the base type for most of the other kinds of objects
+    /// defined in the Activity Vocabulary, including other Core types such as Activity, IntransitiveActivity,
+    /// Collection and OrderedCollection. 
+    /// </summary>
+    public LinkableList<ASObject>? Object
```
[the remark this answers is not in the document]
Comment (warriordog) [reply Source/ActivityPub.Types/AS/ASActivity.cs:80]: flattened back into ASActivity

File: Source/ActivityPub.Types/AS/ASActivity.cs (hunk 1 above)
Comment (warriordog) [reply Source/ActivityPub.Types/AS/ASActivity.cs:117]: flattened in ASActivity

File: Test/ActivityPub.Types.Tests/Integration/Serialization/ComplexObjectSerializationTests.cs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            Target = new LinkableList<ASObject>(collection)
+        };
+        
+        JsonUnderTest.Should().HaveStringProperty("type", "Add");
+        JsonUnderTest.Should().HaveProperty("object");
+        JsonUnderTest.GetProperty("object").Should().HaveStringProperty("type", "Note");
+        // Fails here, Collections are serializing as an array. Should be an object with an items property.
+        JsonUnderTest.GetProperty("target").Should().HaveStringProperty("type", "Collection");
+        JsonUnderTest.GetProperty("target").Should().HaveStringProperty("id", "https://home.example/actor/collections/1");
+        JsonUnderTest.GetProperty("target").GetProperty("items").Should().BeJsonObject(); // Really an array
```
Comment (jenniferplusplus) [inline Test/ActivityPub.Types.Tests/Integration/Serialization/ComplexObjectSerializationTests.cs:88]: Collections don't seem to be serializing properly. This test object serializes as
```
{
  "type": "Add",
  "@context": "https://www.w3.org/ns/activitystreams",
  "actor": "https://home.example/actor",
  "object": {
    "type": "Note",
    "@context": "https://www.w3.org/ns/activitystreams",
    "id": "https://peer.example/actor/activities/1",
    "content": "This is a note"
  },
  "target": [
    "https://example.com/whatever",
    "https://example.com/whatever-2"
  ]
}
```

I expected this
```
{
  "type": "Add",
  "@context": "https://www.w3.org/ns/activitystreams",
  "actor": "https://home.example/actor",
  "object": {
    "type": "Note",
    "@context": "https://www.w3.org/ns/activitystreams",
    "id": "https://peer.example/actor/activities/1",
    "content": "This is a note"
  },
  "target": {
    "items": [
      "https://example.com/whatever",
      "https://example.com/whatever-2"
    ],
    "id": "https://home.example/actor/collections/1",
    "type": "Collection"
  }
}
```

File: Test/ActivityPub.Types.Tests/Integration/Deserialization/MastodonInteropTests.cs [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+                          "object": {
+                            "atomUri": "https://mastodon.example/users/john/statuses/9876",
+                            "attachment": [],
+                            "attributedTo": "https://mastodon.example/users/john",
+                            "cc": ["https://mastodon.example/users/john/followers"],
+                            "content": "<p>Creating test data 🆕</p>",
+                            "contentMap": {
+                              "en": "<p>Creating test data 🆕</p>",
+                              "es": "<p>Creando datos de prueba 🆕</p>"
+                            },
```
Comment (jenniferplusplus) [inline Test/ActivityPub.Types.Tests/Integration/Deserialization/MastodonInteropTests.cs:93]: I'm not sure if the `Content` property is supposed to be picking up `ContentMap` or not. But, it's not in this case

**Comment by jenniferplusplus** [comment]
@warriordog I skipped the failing tests, because I think they're useful examples. I think we can merge, if you're ready.

File: Test/ActivityPub.Types.Tests/Integration/Serialization/ComplexObjectSerializationTests.cs (hunk 4 above)
Comment (warriordog) [reply Test/ActivityPub.Types.Tests/Integration/Serialization/ComplexObjectSerializationTests.cs:88]: To get the expected output, the property type will need to be ASCollection instead of LinkableList. Or, I suppose you could put a collection in the list, and then put items in the collection. What does the spec say about this? If it should be a collection, then let's change the property type. But if this is a Mastodon Special, then maybe the collection-in-a-list would be better?

Comment (jenniferplusplus) [reply Test/ActivityPub.Types.Tests/Integration/Serialization/ComplexObjectSerializationTests.cs:88]: The spec doesn't say anything special about the Target property. I've gotten in the habit of assuming that everything that can be a list is a list, because it removes a lot of edge cases. But maybe I shouldn't be?

Comment (warriordog) [reply Test/ActivityPub.Types.Tests/Integration/Serialization/ComplexObjectSerializationTests.cs:88]: I think it can be a list (because pretty much everything can be), but if it's a list then there should *not* be a `Collection` object. Lists map directly to JSON arrays, while collections have the whole object thing.

Comment (warriordog) [reply Test/ActivityPub.Types.Tests/Integration/Serialization/ComplexObjectSerializationTests.cs:88]: ohhh nevermind I read the code backwards :woman_facepalming:

Comment (warriordog) [reply Test/ActivityPub.Types.Tests/Integration/Serialization/ComplexObjectSerializationTests.cs:88]: I was looking at `Object` - which has no collection - instead of `Target` - which does. I'll take a look, there should definitely be an object there. I'm pretty sure I know the issue too.

File: Test/ActivityPub.Types.Tests/Integration/Deserialization/MastodonInteropTests.cs (hunk 5 above)
Comment (warriordog) [reply Test/ActivityPub.Types.Tests/Integration/Deserialization/MastodonInteropTests.cs:93]: This is an oversight on my part. I've documented it in #12 to be fixed ASAP.

**Review by warriordog — APPROVED** [review]
looks good! thank you for the contribution, and I appreciate your help with all the testing and debugging!
